src/main/java/org/reactome/addlinks/referencecreators/EnsemblBioMartMicroarrayPopulator.java

```diff
@@ -0,0 +1,159 @@
+     * This is to handle cases where a mapping comes from a third-party file that may contain source identifiers that are in *their* system but aren't actually in Reactome.
+     * @throws Exception - Thrown by the MySQLAdaptor/GKInstance classes. IOException/ParseException can also be thrown when getting species names from Species.json file.
+     */
+    @Override
+    public void createIdentifiers(long personID, Map<String, Map<String, List<String>>> mappings, List<GKInstance> sourceReferences) throws Exception {
+
+        // Iterate through each species that we add microarray data to.
+        for (String speciesName : EnsemblBioMartUtil.getSpeciesNames()) {
+            logger.info("Populating microarray data for " + speciesName);
+            String speciesBiomartName = EnsemblBioMartUtil.getBioMartSpeciesName(speciesName);
+
+            // Retrieve identifier mappings that are used from the 'super' mapping.
+            // TODO: Helper class for getting species data
+            Map<String, List<String>> proteinToTranscripts = EnsemblBioMartUtil.getProteinToTranscriptsMappings(speciesBiomartName, mappings);
+
+            Map<String, List<String>> transcriptToProbes = EnsemblBioMartUtil.getTranscriptToMicroarraysMappings(speciesBiomartName, mappings);
+
+            Map<String, List<String>> uniprotToProtein = EnsemblBioMartUtil.getUniprotToProteinsMappings(speciesBiomartName, mappings);
+
+            // We need Uniprot identifiers to retrieve corresponding ReferenceGeneProduct instances (RGP) from the DB,
+            // Ensembl Protein identifiers to retrieve Ensembl Transcript identifiers, and Transcript identifiers
+            // to retrieve Microarray identifiers, which are inserted into the 'otherIdentifier' attribute of the RGP.
+            if (allDataStructuresPopulated(proteinToTranscripts, transcriptToProbes, uniprotToProtein)) {
+                // Retrieve Reactome Species instance and all RGPs associated with it.
+                GKInstance speciesInst = (GKInstance) adapter.fetchInstanceByAttribute(ReactomeJavaConstants.Species, ReactomeJavaConstants.name, "=", speciesName).iterator().next();
+                Collection<GKInstance> rgpInstances = adapter.fetchInstanceByAttribute(ReactomeJavaConstants.ReferenceGeneProduct, ReactomeJavaConstants.species, "=", speciesInst);
+                // Get all RGP identifiers in database, mapping them to associated instances.
+                Map<String, List<GKInstance>> rgpIdentifiersToRGPs = mapRGPIdentifiersToRGPInstances(rgpInstances);
+                // Iterate through each identifier, and then through each RGP instance.
+                for (String rgpIdentifier : rgpIdentifiersToRGPs.keySet()) {
+                    for (GKInstance rgpInst : rgpIdentifiersToRGPs.get(rgpIdentifier)) {
+                        addMicroarrayDataToReferenceGeneProduct(rgpIdentifier, rgpInst, speciesBiomartName, mappings);
+                    }
+                }
```

Comment:
I'm a bit confused about why mapping the RGP identifiers to their instances is needed.  Can a collection of RGP instances be used instead since the identifier can always be queried from the instance?

Reply:
This was how the Perl code did it. Early on in the code I just aped the Perl code, and this part didn't end up needing to be touched again (to make the data population work). I'll look into doing it this way

Reply:
Made the adjustment. Much cleaner!